### PR TITLE
RFC3339 adapter.

### PR DESCRIPTION
**Pull request by swankjesse**
Much thanks to Jackson for doing all the real work.

### CODE REVIEW

File: adapters/src/main/java/com/squareup/moshi/Iso8601Utils.java [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+
+  /** The GMT timezone, prefetched to avoid more lookups. */
+  static final TimeZone TIMEZONE_Z = TimeZone.getTimeZone(GMT_ID);
+
+  /** Returns {@code date} formatted as yyyy-MM-ddThh:mm:ss.sssZ */
+  public static String format(Date date) {
+    Calendar calendar = new GregorianCalendar(TIMEZONE_Z, Locale.US);
+    calendar.setTime(date);
+
+    // estimate capacity of buffer as close as we can (yeah, that's pedantic ;)
```
Comment (JakeWharton) [inline adapters/src/main/java/com/squareup/moshi/Iso8601Utils.java]: https://xkcd.com/541/

File: adapters/src/test/java/com/squareup/moshi/Rfc3339DateJsonAdapterTest.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+  @Test public void fromJson() throws Exception {
+    assertThat(adapter.fromJson("\"1970-01-01T00:00:00.000Z\""))
+        .isEqualTo(newDate(1970, 1, 1, 0, 0, 0, 0, 0));
+    assertThat(adapter.fromJson("\"1985-04-12T23:20:50.520Z\""))
+        .isEqualTo(newDate(1985, 4, 12, 23, 20, 50, 520, 0));
+    assertThat(adapter.fromJson("\"1996-12-19T16:39:57-08:00\""))
+        .isEqualTo(newDate(1996, 12, 19, 16, 39, 57, 0, -8 * 60));
+    assertThat(adapter.fromJson("\"1990-12-31T23:59:60Z\""))
+        .isEqualTo(newDate(1990, 12, 31, 23, 59, 59, 0, 0));
```
Comment (swankjesse) [inline adapters/src/test/java/com/squareup/moshi/Rfc3339DateJsonAdapterTest.java]: truncated leap seconds!

File: adapters/src/main/java/com/squareup/moshi/Iso8601Utils.java [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+      throw new NumberFormatException(value);
+    }
+    // use same logic as in Integer.parseInt() but less generic we're not supporting negative values
+    int i = beginIndex;
+    int result = 0;
+    int digit;
+    if (i < endIndex) {
+      digit = Character.digit(value.charAt(i++), 10);
+      if (digit < 0) {
+        throw new NumberFormatException("Invalid number: " + value);
```
Comment (JakeWharton) [inline adapters/src/main/java/com/squareup/moshi/Iso8601Utils.java]: This should be `value.substring(beginIndex, endIndex)`

File: adapters/src/main/java/com/squareup/moshi/Iso8601Utils.java [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+      }
+      result = -digit;
+    }
+    while (i < endIndex) {
+      digit = Character.digit(value.charAt(i++), 10);
+      if (digit < 0) {
+        throw new NumberFormatException("Invalid number: " + value);
```
Comment (JakeWharton) [inline adapters/src/main/java/com/squareup/moshi/Iso8601Utils.java]: Similarly

**Comment by JakeWharton** [comment]
LGTM

File: adapters/src/main/java/com/squareup/moshi/Iso8601Utils.java (hunk 4 above)
Comment (swankjesse) [inline adapters/src/main/java/com/squareup/moshi/Iso8601Utils.java]: https://github.com/FasterXML/jackson-databind/pull/949

File: adapters/src/main/java/com/squareup/moshi/Iso8601Utils.java (hunk 3 above)
Comment (swankjesse) [inline adapters/src/main/java/com/squareup/moshi/Iso8601Utils.java]: https://github.com/FasterXML/jackson-databind/pull/949

**Pull request merged**